examples/sampling/tool_use.py

```
@@ -0,0 +1,122 @@
+        result = await super().__call__(messages, params, context)
+        console.print("      [bold blue]SAMPLING[/] Response received")
+        return result
+
+
+# Define tools available to the LLM during sampling
+def calculate(expression: str) -> str:
+    """Evaluate a mathematical expression. Example: '2 + 2' or '10 * 5'"""
+    console.print(f"        [bold magenta]TOOL[/] calculate({expression})")
+    try:
+        # Safe eval for simple math
+        result = eval(expression, {"__builtins__": {}}, {})
```

Ruff (0.14.8)

Use of possibly insecure function; consider using `ast.literal_eval`

(S307)